// flat-ecs/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var World_1 = require("./World");
Object.defineProperty(exports, "World", { enumerable: true, get: function () { return World_1.default; } });
var Archetype_1 = require("./Archetype");
Object.defineProperty(exports, "Archetype", { enumerable: true, get: function () { return Archetype_1.Archetype; } });
var Component_1 = require("./Component");
Object.defineProperty(exports, "Component", { enumerable: true, get: function () { return Component_1.Component; } });
var ComponentMapper_1 = require("./ComponentMapper");
Object.defineProperty(exports, "ComponentMapper", { enumerable: true, get: function () { return ComponentMapper_1.ComponentMapper; } });
var EntityIteratingSystem_1 = require("./EntityIteratingSystem");
Object.defineProperty(exports, "EntityIteratingSystem", { enumerable: true, get: function () { return EntityIteratingSystem_1.EntityIteratingSystem; } });
var EntitySystem_1 = require("./EntitySystem");
Object.defineProperty(exports, "EntitySystem", { enumerable: true, get: function () { return EntitySystem_1.EntitySystem; } });
var System_1 = require("./System");
Object.defineProperty(exports, "System", { enumerable: true, get: function () { return System_1.System; } });
var InjectMapper_1 = require("./InjectMapper");
Object.defineProperty(exports, "InjectMapper", { enumerable: true, get: function () { return InjectMapper_1.InjectMapper; } });
var Inject_1 = require("./Inject");
Object.defineProperty(exports, "Inject", { enumerable: true, get: function () { return Inject_1.Inject; } });
